.changeset/rotten-colts-decide.md

```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-catalog-backend-module-incremental-ingestion': minor
+---
+
+Allow incremental event handlers to be async; Force event handler to indicate if it made a change
```

Comment:
Maybe you can add some instructions here as to how to fix the breaking changes? :pray:

Reply:
Ok, I'll add that. I could also make it backwards compatible so that if it is `undefined`, we treat it as an ignore and issue a warning.

Reply:
Yeah think that would be nice - wonder if it also makes sense to have backwards compatibility for the promise to so you can provide either `ReturnType | Promise<ReturnType>`?

Reply:
@cowboyd ^

Reply:
My preference would be to keep the return type as simple as possible. Also, `await`ing a value returns that value.

Reply:
`await`ing is not the issue here, the issue is that it's a breaking change for the input types right as you now NEED to provide a promise is what I'm getting at. 

What I mean is you could do `onEvent: (params: EventParams) => Promise<IncrementalEntityEventResult> | IncrementalEntityEventResult;` right? So that you can still proving a sync function.

We're breaking the API so it doesn't really matter either way. So maybe it doesn't really matter.

Reply:
It will break the TypeScript build, but it won't actually break the code at runtime because whether the handler returns a promise or a simple value, it will await it, and get the desired value.

https://github.com/thefrontside/backstage/blob/incremental-ingestion-async-event-hook/plugins/catalog-backend-module-incremental-ingestion/src/engine/IncrementalIngestionEngine.ts#L350

That said, I'd say go ahead and break the API now, in a pre 1.0 version. After all, isn't that what pre 1.0 versions are for?